editor/src/layout/widgets.rs

```diff
@@ -22,6 +22,149 @@ pub trait PropertyHolder {
+		match self {
+			Layout::MenuLayout(menu_layout) => Box::new(menu_layout.iter_mut()),
+			Layout::WidgetLayout(widget_layout) => Box::new(widget_layout.iter_mut()),
+		}
+	}
+}
+
+impl Default for Layout {
+	fn default() -> Self {
+		Layout::WidgetLayout(WidgetLayout::default())
+	}
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct MenuEntry {
+	pub label: String,
+	pub icon: Option<String>,
+	pub children: Option<Vec<Vec<MenuEntry>>>,
+	pub action: WidgetHolder,
+	pub shortcut: Option<Vec<Key>>,
+}
+
+impl MenuEntry {
+	pub fn create_action(callback: impl Fn(&()) -> Message + 'static) -> WidgetHolder {
+		WidgetHolder::new(Widget::Invisible(Invisible {
+			on_update: WidgetCallback::new(callback),
+		}))
+	}
+
+	pub fn no_action() -> WidgetHolder {
+		MenuEntry::create_action(|_| Message::NoOp)
+	}
+}
+
+impl Default for MenuEntry {
+	fn default() -> Self {
+		Self {
+			action: MenuEntry::create_action(|_| DialogMessage::RequestComingSoonDialog { issue: None }.into()),
+			label: "".into(),
+			icon: None,
+			children: None,
+			shortcut: None,
+		}
+	}
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct MenuColumn {
+	pub label: String,
+	pub children: Vec<Vec<MenuEntry>>,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct MenuLayout {
+	pub layout: Vec<MenuColumn>,
+}
+
+impl MenuLayout {
+	pub fn new(layout: Vec<MenuColumn>) -> Self {
+		Self { layout }
+	}
+
+	pub fn iter(&self) -> impl Iterator<Item = &WidgetHolder> + '_ {
+		MenuLayoutIter {
+			stack: self.layout.iter().flat_map(|column| column.children.iter()).flat_map(|group| group.iter()).collect(),
+		}
+	}
+
+	pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut WidgetHolder> + '_ {
+		MenuLayoutIterMut {
+			stack: self.layout.iter_mut().flat_map(|column| column.children.iter_mut()).flat_map(|group| group.iter_mut()).collect(),
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+pub struct MenuLayoutIter<'a> {
+	pub stack: Vec<&'a MenuEntry>,
+}
+
+impl<'a> Iterator for MenuLayoutIter<'a> {
+	type Item = &'a WidgetHolder;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		match self.stack.pop() {
+			Some(menu_entry) => {
+				self.stack.extend(menu_entry.children.iter().flat_map(|group| group.iter()).flat_map(|entry| entry.iter()));
+				Some(&menu_entry.action)
+			}
+			None => None,
+		}
+	}
+}
+
+pub struct MenuLayoutIterMut<'a> {
+	pub stack: Vec<&'a mut MenuEntry>,
+}
+
+impl<'a> Iterator for MenuLayoutIterMut<'a> {
+	type Item = &'a mut WidgetHolder;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		match self.stack.pop() {
+			Some(menu_entry) => {
+				self.stack.extend(menu_entry.children.iter_mut().flat_map(|group| group.iter_mut()).flat_map(|entry| entry.iter_mut()));
+				Some(&mut menu_entry.action)
+			}
+			None => None,
+		}
+	}
+}
+
 #[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
 pub struct WidgetLayout {
 	pub layout: SubLayout,
```

Comment:
(This comment applies ~27 lines below the targeted line since GitHub doesn't let me comment on that.)

Are you able to address the `// TODO: Rename LayoutRow to something more generic` in this PR? It seems pretty related to the work you're doing and you are probably better able to figure out a name for this than me with your understanding of the full system design.

Reply:
I renamed it to `LayoutGroup`. What do you think?

Reply:
Good name!